src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -644,6 +643,60 @@ void FileSegment::completePartAndResetDownloader()
     LOG_TEST(log, "Complete batch. ({})", getInfoForLogUnlocked(lk));
 }
 
+void FileSegment::shrinkFileSegmentToDownloadedSize(const LockedKey & locked_key, const FileSegmentGuard::Lock & lock)
+{
+    chassert(downloaded_size);
+    chassert(fs::file_size(getPath()) > 0);
+
+    if (downloaded_size == range().size())
+    {
+        /// Nothing to resize;
+        return;
+    }
+
+    if (!locked_key.isLastOwnerOfFileSegment(offset()))
+    {
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Shrinking of file segment can  be done only by the last holder: {}",
+            getInfoForLog());
+    }
+
+    size_t result_size = downloaded_size;
+
+    const bool shrink_to_aligned_size = !created_from_write_through_cache;
```

Comment:
maybe then `shrink_to_aligned_size` would be a better name for `CreateFileSegmentSettings` parameter?